File: Ml/database/d5.py
import pymysql

# 함수화
def selectData( uid ):
  connection = None
  result     = None
  # 아이디가 guest인 데이터만 가져와라
  try:
    connection = pymysql.connect(host='localhost',
                                user='root',
                                password='12341234',
                                db='python_db',
                                charset='utf8mb4'                              
                                ,cursorclass=pymysql.cursors.DictCursor)
    with connection.cursor() as cursor:     
      # 파라미터를 넣을때는 '값' => %s로 치환(홀따움표 포함)
      sql = '''
      select src, slang as label from tbl_trans_log
      where uid=%s;
      '''
      # 파리미터는 쿼리 수행시 튜플로 전달(2번인자)
      cursor.execute(sql, (uid,))
      result = cursor.fetchall() 
      #print(result)    
  except Exception as e:
    print('오류발생', e)
  finally:
    if connection:
      connection.close()
  return result

# 번역 요청 데이터 삽입하는 함수
def insertData( src, out, slang, olang, uid='guest' ):
  connection = None
  result     = 0   
  try:
    connection = pymysql.connect(host='localhost',
                                user='root',
                                password='12341234',
                                db='python_db',
                                charset='utf8mb4'                              
                                ,cursorclass=pymysql.cursors.DictCursor)
    with connection.cursor() as cursor:           
      sql = '''
      INSERT INTO `tbl_trans_log` 
        (`src`, `out`, `slang`, `olang`, `uid`) 
      VALUES 
        (%s, %s, %s, %s, %s);
      '''      
      cursor.execute(sql, (src, out, slang, olang, uid))
    # 데이터베이스에 실제 반영
    connection.commit()
    # 영향을 받은 로의수
    result = connection.affected_rows()
  except Exception as e:
    print('오류발생', e)
  finally:
    if connection:
      connection.close()
  # 결과 리턴
  return result

# 테스트 코드
# 모듈화를 시켜서 다른 모듈이 selectData()를 사용할려고, 모듈가져오기를 수행
# 아래 코드가 수행됨
if __name__ == '__main__': # 위의 문제 해결
  print( '->', selectData('guest') )
  print( '->', insertData('Helloworld', '안녕', 'en', 'ko') )

File: Another_Web/node_basic/database/d5.js
// 해당 코드를 누구나 불러서 사용할수 있게 함수의 모듈화를 구성
const mysql = require('mysql')

let selectLogin = ( uid, upw, cb )=>{
    var connection = mysql.createConnection({
        host     : 'localhost',
        user     : 'root',
        password : '12341234',
        database : 'nodedb'
    });
    connection.connect( (err)=>{
        if( err ){
            //console.log('접속 오류', err)
            // 접속 오류가 발생하면 콜백을 호출하여 해당 상황을 전달
            cb( err )
        }else{   
            let sql = `select * from tbl_users where uid = ? and upw = ? ;`
            connection.query( sql, [uid, upw], (error, rows)=>{                        
                connection.end();
                // 결과를 돌려줘라 2
                cb( error, rows )
            } ) 
        }
    } );
}

selectLogin( 'guest', '1', ( err, rows )=>{
    if(err){
        // 접속 오류 or 쿼리 오류 등등 오류가 발생한 상황
        console.log('접속오류 or 쿼리 오류')
    }else{
        // 쿼리 성공후 결과를 정상적으로 받은 상황
        //console.log( rows )        
        if( rows.length == 0 ){
            // 회원아님
            console.log( '회원아님' )
        }else{
            // 배열의 길이가 0보다 크면 회원이다
            console.log( '회원임', rows[0].name+'님 방갑습니다' )
        }
    }
} )